design/design-doc-integration-tests.md

```diff
@@ -0,0 +1,298 @@
+Everything that happens between the start and stop of the controller manager
+should be handled in the test implementations themselves.
+
+### System Design
+
+The framework encompasses three broad responsibilities:
+
+1. Environment Setup
+
+The framework sets up an environment by taking a Kubernetes REST configuration
+and installing CustomResourceDefinitions, starting a controller manager, and
+creating a client that can be used to communicate to the cluster. Because the
+framework is a wrapper around the [envtest] package from [controller-runtime] it
+can also start a local control plane if no REST configuration is applied.
+Minimal setup that just installs CRDs into an existing cluster and returns a
+controller manager would look as follows:
+
+```go
+cfg, err := config.GetConfig()
+if err != nil {
+    t.Fatal(err)
+}
+
+a, err := athodyd.New(cfg, athodyd.WithCRDDirectoryPaths([]string{"../crds"}))
+if err != nil {
+    t.Fatal(err)
+}
+```
+
+The return value of `athodyd.New()` is of type `*athodyd.Manager`, which
+implements the [controller-runtime] `manager.Manager` interface. This means that
+controllers and API types can be added to the manager in the same manner they
+are when the controller is being run in production:
+
+```go
+addToScheme(a.GetScheme()) // add API types to the manager scheme
+controllerSetupWithManager(a) // register controllers with the manager
+```
+
+Once all setup is complete, the manager can be started with `a.Run()`.
+
+2. Environment Connection
+
+It is expected that most integration tests will want to interact with the API
+server. To do so, a client can be retrieved and injected into the test function
+at runtime.
+
+```go
+for name, tc := range cases {
+    t.Run(name, func(t *testing.T) {
+        err := tc.test(a.GetClient()) // retrieve the kubernetes client from athodyd
+        if err != nil {
+            t.Error(err)
+        }
+    })
+}
+```
+
+3. Environment Cleanup
+
+Because testing controllers usually involves creating some number of CRDs, it is
+necessary to perform cleanup when external clusters are used. The framework will
+default to deleting all CRDs that were installed in environment setup during its
+cleanup if no alternative cleanup is supplied. During this step, the controller
+manager is stopped, the `Cleaner` function is executed, and the connection to
+the cluster will be terminated. If no configuration was supplied in the initial
+setup, the local control plane will be destroyed in this step as well. If your
+tests are dependent on successful cleanup, it may be desirable to fail if an
+error is returned:
+
+```go
+defer func() {
+    if err := a.Cleanup(); err != nil {
+        t.Fatal(err)
+    }
+}()
+```
+
+To override the default `Cleaner` function, pass in your own to `athodyd.New()`:
+
+```go
+a, err := athodyd.New(cfg,
+    athodyd.WithCRDDirectoryPaths([]string{"../crds"}),
+    athodyd.WithCleaner(func(*envtest.Environment, client.Client){ return nil }))
```

Comment:
What do you think about:

1. Either making this `WithCleaners(cleaner ...OperationFn)` or introducing an `OperationChain` type that chains operations.
1.  Renaming [`athodyd.NewCleaner`](https://github.com/hasheddan/athodyd/blob/5821b7d750d33c9cdd0a38581f97ab2adbba5ca4/athodyd.go#L58) to `athodyd.NewCRDCleaner`

I think most folks are going to want to clean their CRDs after running a test, so it would be nice to be able to write something like `athodyd.New(cfg, athodyd.Cleaners(myCustomCleaner, athodyd.NewCRDCleaner()))`.